### PR TITLE
Ticket3562 Added upgrade for banner.xml customisation

### PR DESCRIPTION
**Pull request by adamb56789**
https://github.com/ISISComputingGroup/IBEX/issues/3562

Changes `banner.xml` to the new format and adds the motor status display and stop motors button which are being moved out of the GUI. If no `banner.xml` exists, creates one and populates it with a default.

See [ticket](https://github.com/ISISComputingGroup/IBEX/issues/3562) for acceptance criteria

### CODE REVIEW

File: src/upgrade_step_from_5p4p1.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                lines[1] = line_2
+
+                # Insert the xml for the motor status and button
+                lines.insert(3, motor_xml)
+
+                with open(path, 'w') as file:
+                    file.writelines(lines)
+
+            except IOError:
+                # In case there isn't a banner xml
```
Comment (Tom-Willemsen) [inline src/upgrade_step_from_5p4p1.py]: This isn't the only possible cause of an `IOError`. If we had some other error we would end up overwriting the banner completely here.

Make this condition `not os.path.exists` or similar

File: src/upgrade_step_from_5p4p1.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            line_2 = '<banner xmlns="http://epics.isis.rl.ac.uk/schema/banner/1.0" xmlns:blk="http://epics.isis.rl.ac.uk/schema/banner/1.0" xmlns:xi="http://www.w3.org/2001/XInclude">\n'
+            motor_xml = '    <item>\n      <button>\n        <name>Stop All Motors</name>\n        <pv>CS:MOT:STOP:ALL</pv>\n        <local>true</local>\n        <pvValue>1</pvValue>\n        <textColour>#000000</textColour>\n        <buttonColour>#e0e0e0</buttonColour>\n        <fontSize>9</fontSize>\n        <width>100</width>\n        <height>25</height>\n      </button>\n    </item>\n    <item>\n      <display>\n        <name>Motors are</name>\n        <pv>CS:MOT:MOVING:STR</pv>\n        <local>true</local>\n        <width>170</width>\n      </display>\n    </item>\n'
```
[the remark this answers is not in the document]
Comment (adamb56789) [reply src/upgrade_step_from_5p4p1.py]: I've changed it to using a triple-quoted string. To keep the formatting nice at the end I also had to remove the spaces and new lines.

File: src/upgrade_step_from_5p4p1.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                for items in tree.getroot():
+                    for item in items:
+                        new_item = ET.SubElement(new_items, "item")
+                        new_display = ET.SubElement(new_item, "display")
+                        for element in item:
+                            ET.SubElement(new_display, element.tag).text = element.text
+                        ET.SubElement(new_display, "width").text = "250"
+
+                # Get pretty xml using minidom and write to file
+                xmlstr = minidom.parseString(ET.tostring(new_root)).toprettyxml(indent="  ")
```
[the remark this answers is not in the document]
Comment (adamb56789) [reply src/upgrade_step_from_5p4p1.py]: elementree doesn't have a good way to pretty print, so I used minidom instead. https://stackoverflow.com/questions/17402323/use-xml-etree-elementtree-to-print-nicely-formatted-xml-files/17402424

File: src/upgrade_step_from_5p4p1.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                with open(path, "w") as f:
+                    f.write(xmlstr)
+
+                with open(path, 'r') as f:
+                    lines = f.readlines()
+
+                # Replace line 2 (not sure if these namespaces are used for anything, except the first, but keeping them anyway)
+                lines[1] = line_2
```
[the remark this answers is not in the document]
Comment (adamb56789) [reply src/upgrade_step_from_5p4p1.py]: I managed to do this more properly by setting attributes of the root node.

File: src/upgrade_step_from_5p4p1.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+                # Insert the xml for the motor status and button
+                lines.insert(3, motor_xml)
```
[the remark this answers is not in the document]
Comment (adamb56789) [reply src/upgrade_step_from_5p4p1.py]: It now parses the XML for the motor status and button with `ET.fromstring` and appends in the right place.

File: src/upgrade_step_from_5p4p1.py (hunk 1 above)
Comment (adamb56789) [reply src/upgrade_step_from_5p4p1.py]: It now uses `if not os.path.exists(banner_path)` instead.

**Comment by adamb56789** [comment]
The broken XML that you have seems to be the result of the script being run multiple times on the same file. I've added a check which looks for a `button` and skips the upgrade if it finds one.

File: src/upgrade_step_from_5p4p1.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                motor_button.replace(" ", "").replace("\n", "").replace("StopAllMotors", "Stop All Motors"))
+            motor_display = ET.fromstring(
+                motor_display.replace(" ", "").replace("\n", "").replace("Motorsare", "Motors are"))
+            config_display = ET.fromstring(config_display.replace(" ", "").replace("\n", ""))
+            if not os.path.exists(banner_path):
+                # In case there isn't a banner xml
+                print("banner.xml not found at " + banner_path + ", making a new banner.xml")
+                with open(banner_path, "w") as f:
+                    f.write(complete_new)
+            elif ET.parse(banner_path).find(".//{http://epics.isis.rl.ac.uk/schema/banner/1.0}button") is not None:
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply src/upgrade_step_from_5p4p1.py]: Changed so we now check if the banner.xml is already valid against the new schema and do not do the upgrade if it is.

We now also check whether the banner.xml is valid against the old schema, if it is not we throw an exception, print to the user why we cannot continue and stop the upgrade.

**Pull request merged**